stdlib/dynarray.mli

```diff
@@ -479,3 +479,121 @@ val reset : 'a t -> unit
     elements.)
 *)
 
+
+
+(** {1:examples Code examples}
+
+{2 Binary heaps for mutable priority queues}
+
+We can use dynamic arrays to implement a mutable priority
+queue. A priority queue provides a function to add elements, and
+a function to extract the minimum element -- according to some
+comparison function.
+
+{[
+(* We present our priority queues as a functor
+   parametrized on the comparison function. *)
+module Heap (Elem : Map.OrderedType) : sig
+  type t
+  val create : unit -> t
+  val add : t -> Elem.t -> unit
+  val pop_min : t -> Elem.t option
+end = struct
+
+  (* Our priority queues are implemented using the standard "min heap"
+     data structure, a dynamic array representing a binary tree. *)
+  type t = Elem.t Dynarray.t let create = Dynarray.create
+
+ (* The node of index [i] has as children the nodes of index [2 * i + 1]
+    and [2 * i + 2] -- if they are valid indices in the dynarray. *)
+  let left_child i = 2 * i + 1
+  let right_child i = 2 * i + 2
+  let parent_node i = (i - 1) / 2
+
+  (* We use indexing operators for convenient notations. *)
+  let ( .%() ) = Dynarray.get
+  let ( .%()<- ) = Dynarray.set
+
+  (* Auxiliary functions to compare and swap two elements
+     in the dynamic array. *)
+  let order h i j =
+    Elem.compare h.%(i) h.%(j)
+
+  let swap h i j =
+    let v = h.%(i) in
+    h.%(i) <- h.%(j);
+    h.%(j) <- v
+
+  (* We say that a heap respects the "heap ordering" if the value of
+     each node is smaller than the value of its children. The
+     algorithm manipulates arrays that respect the heap algorithm,
+     except for one node whose value may be too small or too large.
+
+     The auxiliary functions [heap_up] and [heap_down] take
+     such a misplaced value, and move it "up" (respectively: "down")
+     the tree by permuting it with its parent value (respectively:
+     a children's value) until the heap ordering is restored. *)
+
+  let rec heap_up h i =
+    if i = 0 then () else
+    let parent = parent_node i in
+    if order h i parent < 0 then
+      (swap h i parent; heap_up h parent)
+
+  and heap_down h ~len i =
+    let left, right = 2 * i + 1, 2 * i + 2 in
```

Comment:
I think you wanted to use `left_child i, right_child i` otherwise the functions are never used.

Reply:
Indeed I was.

(I wrote the code without them, but then I tried to understand how to do literate-programming explanations of the code inside the example, and I started explaining the indexing structure much earlier, so I moved the computations out as auxiliary functions.)

Thanks!